### PR TITLE
kubeadm etcd liveness probe changes.

### PR DESCRIPTION
**Pull request by mauilion**
**What this PR does / why we need it**:

- This changes the logic in kubeadm alpha phase etcd local to produce a
working url vs an ip for the liveness probe.

- Gets rid of the need for a statically defined port (2379) 

- Moved to --advertise-client-url instead of listen-client-url as the
flag can be set to 0.0.0.0  and doesn't need to match the cert.

- Adds some logic to validate that listen address is a valid etcd url.
If not returns https://127.0.0.1:2379 which should match the default
etcd local setup.

- Modified tests to fit new code and all tests pass with:
make test WHAT=./cmd/kubeadm/app/util/staticpod/ GOFLAGS="-v"

Signed-off-by: Duffie Cooley <dcooley@heptio.com>

**Special notes for your reviewer**:
NONE

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by mauilion** [comment]
@stealthybox Can you take a look?

File: cmd/kubeadm/app/util/staticpod/utils.go
Comment (stealthybox) [inline cmd/kubeadm/app/util/staticpod/utils.go]: This `url.Parse` needs an error handler

given:
```golang
tests := []string{
        "http://[2001:db8::1]:2379",
        "http://domain.com:2379",
        "domain.com",
        "domain",
        "http://:2379",
        "http://2379",
        "hello://2379",
        ":2379",
        "http://",
        "://",
        "blah//",
        "",
}
```
It will fail with these errors and cause the code to panic
```
:2379
error parsing etcdAdvertiseClientURLsArg: parse :2379: missing protocol scheme

://
error parsing etcdAdvertiseClientURLsArg: parse ://: missing protocol scheme
```

I'd probably change it to:
```golang
if u, err := url.Parse(arg); err == nil {
  ...
}
```

Given that the function signature currently does not return an error, it's probably appropriate to return `https://127.0.0.1:2379` in the case of a Parse error.
This is the simplest thing to do, but may not be the best for the user.
It may be more appropriate to fail fatally, but this would likely require work to return an error and add error handling to the caller.

It may be an easy workaround to simply print a warning in the case that the flag fails to parse.
This is less of a big deal if there are already other functions that are implementing this workaround.

Comment (stealthybox) [reply cmd/kubeadm/app/util/staticpod/utils.go]: It might be good to maintain support for `etcdListenClientURLsArg` as a fallback to `etcdAdvertiseClientURLsArg` so that we don't unnecessarily deprecate that functionality.
This would involve un-removing that code into this new function.

What do you think?

File: cmd/kubeadm/app/util/staticpod/utils.go
Comment (stealthybox) [inline cmd/kubeadm/app/util/staticpod/utils.go]: Thanks for providing this link.
Would you please pin it to the commit?
https://github.com/coreos/etcd/blob/54d56e2531a3d3fff0f80a15371cc18a68b83e52/pkg/types/urls.go#L35-L47

**Review by stealthybox — CHANGES_REQUESTED** [review]
@mauilion: Thanks so much for taking the time to make this use-case work ✨

Here are some comments from my review over the arctic circle :)
✈️🇦🇶

File: cmd/kubeadm/app/util/staticpod/utils.go
Comment (stealthybox) [reply cmd/kubeadm/app/util/staticpod/utils.go]: ( oh, it would be good to add tests for ipv6 urls as well as the invalid panic strings. )

**Review by timothysc — APPROVED** [review]
/approve  

@stealthybox and @kubernetes/sig-cluster-lifecycle-pr-reviews for lgtm

File: cmd/kubeadm/app/util/staticpod/utils.go
Comment (mauilion) [reply cmd/kubeadm/app/util/staticpod/utils.go]: @stealthybox What do you think of something like this for the error handler part: https://play.golang.org/p/PEIQegS9fET

**Review by detiber — COMMENTED** [review]
This lgtm once the comments from @stealthybox have been addressed.

**Comment by luxas** [comment]
@timothysc I suppose we want to get this into v1.11?

**Comment by timothysc** [comment]
@luxas yup.

**Comment by luxas** [comment]
Discussed with @timothysc now and we've changed our minds, we **won't** merge this for v1.11 to lower the risk. Happy to merge something like this later when we have more data on all the edge cases, and the PR is reworked as @stealthybox said.

**Comment by stealthybox** [comment]
@mauilion I implemented the simple solution we discussed:
https://github.com/kubernetes/kubernetes/pull/64670

PTAL when you can :+1:

/hold

**Pull request closed without merging**
